[Feat] 케어콜 푸시알림 설정 #110

alswlekk opened the pull request:
## 🔗 관련 이슈
<!-- 이 PR과 연결된 이슈 번호를 명시해주세요 (예: Close #123) -->
- Close #110

## 📙 작업 설명
<!-- 주요 수정 사항이나 개발 내용을 요약해주세요 -->
- 케어콜과 관련해서 FCM을 통한 알림 구현 코드를 추가했습니다.
- 앱 내에서 알림 아이콘은 단색으로만 구현해야해서 일단 메디의 몸통으로 뜨도록 하고, 앱 밖에선 앱의 아이콘으로 상단 표시줄에 뜨게 됩니다.
다음 문서들을 참고해 코드 작성했습니다
https://firebase.google.com/docs/cloud-messaging/get-started?platform=android&hl=ko
https://like-while.tistory.com/65
추가적으로 SharedPerferences를 이용하는 방식에서 DataStore에 저장하는 방식으로 수정했습니다

- **암호화 도구** → `FcmTokenEncryptor`
- **DataStore에 저장할 데이터 모델** → `data class FcmToken(...)`
- **DataStore Serializer (읽을 때 복호화 / 쓸 때 암호화)** → `FcmTokenSerializer`
- **레포지토리 인터페이스** → `FcmRepository`
- **레포지토리 구현체** → `FcmRepositoryImpl`

## 📸 스크린샷 또는 시연 영상 (선택)
<!-- UI 변경사항이 있다면 이미지나 GIF를 첨부해주세요 -->
|기능|미리보기|기능|미리보기|
|:--:|:--:|:--:|:--:|
<img width="1080" height="2340" alt="image" src="https://github.com/user-attachments/assets/341ed910-6a3a-47bb-896c-22dbaa8fc54b" />
<img width="1080" height="2340" alt="image" src="https://github.com/user-attachments/assets/d7bb29aa-0211-4bbc-9a7d-b3e504e28e05" />

| 기능 설명 |<img src="링크" width="300" />| 기능 설명 |<img src="링크" width="300" />|

## 💬 추가 설명 or 리뷰 포인트 (선택)
<!-- 리뷰어가 중점적으로 봐야 할 부분이나 설명이 필요한 내용을 자유롭게 작성해주세요 -->
-


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **새 기능**
  * 푸시 알림 수신·표시 기능 추가 및 백그라운드 서비스 등록
  * 앱 시작 시 알림 권한 요청 UI 추가
  * FCM 토큰 암호화 저장 및 자동 동기화/검증 로직 도입

* **변경 사항**
  * 빌드 설정 및 의존성 업데이트(네트워킹, DataStore, Firebase 등)
  * 회원 가입 및 알림 관련 API/페이로드에 FCM 토큰 포함
  * 앱 아이콘·알람 아이콘 리소스 추가 및 CI/코드 스타일 검사 강화
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

alswlekk committed:
Feat: #110 케어콜 푸시알림 설정(잠금화면 시에도 뜨도록)

alswlekk committed:
Merge branch 'develop' of github.com:Medicare-Call/Medicare-Call-Android into feat/fcm-alarm-#110

# Conflicts:
#	app/build.gradle.kts

alswlekk committed:
Merge branch 'develop' into feat/fcm-alarm-#110

alswlekk committed:
Feat: #110 hiltManager 관련 의존성 추가

alswlekk committed:
Feat: build.gradle 파일에 hilt manager 관련 코드 추가

alswlekk committed:
Feat: google-services.json 파일 존재 여부 상관없이 테스트 진행되도록 수정

alswlekk committed:
Refactor Android CI workflow for clarity and structure

alswlekk committed:
Feat: 패키지명 수정

alswlekk committed:
Update package name in android-ci.yml

alswlekk committed:
Update API key in android-ci.yml

alswlekk committed:
Update google-services.json creation in CI workflow

ikseong00 commented on .github/workflows/android-ci.yml:
```diff
@@ -37,13 +37,27 @@ jobs: # CI에서 수행할 작업을 정의한다.
+    "current_key": "DUMMY_API_KEY"
+  }]
+}
+EOF
```
요 부분은 아래 ci 파일 참고해서 작성해주세요!
secret 으로 넘겨야합니다..
https://github.com/OurMenu/OurMenu-Android-V2/blob/develop/.github/workflows/android-ci.yml

ikseong00 reviewed (approved):
레퍼런스 파일 첨부 좋습니다!
ci 파일 체크만 부탁드려요~

alswlekk committed:
Update CI workflow to generate google-services.json

alswlekk committed:
Merge branch 'develop' into feat/fcm-alarm-#110

alswlekk committed:
Refactor Android CI workflow for clarity and updates

alswlekk committed:
Refactor Android CI workflow for google-services.json

Updated the Android CI workflow to generate google-services.json using Base64 decoding and added comments for clarity.

librawish808 commented on app/src/main/java/com/konkuk/medicarecall/data/api/AppPreferences.kt:
```diff
@@ -0,0 +1,40 @@
+class AppPreferences @Inject constructor(
+    @ApplicationContext context: Context,
+) {
+    private val sharedPreferences: SharedPreferences =
```
현재 AppPreferences에 access_token과 fcm_token이 SharedPreferences에 저장되는 걸로 보입니다. 이 부분도 동일하게 저번 PR(#111)처럼 DataStore(+직렬화)로 통일하면 어떨지 의견드립니다!

alswlekk replied:
좋습니다 ! 수정해서 올릴게요 ~

alswlekk replied:
수정했습니다 ~

librawish808 commented on app/src/main/java/com/konkuk/medicarecall/data/api/FcmService.kt:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    private fun showNotification(remoteMessage: RemoteMessage) {
+        val channelId = "fcm_alert"
```
위쪽에서는 App.FCM_CHANNEL_ID 상수를 쓰고, 여기서는 channelId를 직접 쓰고 있는데 통일하면 어떨까요?

alswlekk replied:
넵 !!!

alswlekk replied:
수정했습니다 ~

librawish808 reviewed (commented):
주석 덕분에 코드 흐름을 이해할 수 있었습니다👍

ProtossManse commented:
CI 아직도 실패했다고 뜨는데 혹시 해결 안된걸까요?

지민님이 말씀해주신대로 SharedPreferences -> DataStore로 바꾸는 게 좋을 것 같습니다 !

alswlekk committed:
Merge branch 'develop' into feat/fcm-alarm-#110

alswlekk committed:
Merge branch 'develop' of github.com:Medicare-Call/Medicare-Call-Android into feat/fcm-alarm-#110

alswlekk committed:
Refactor: FCM 관련 코드 리펙토링

alswlekk committed:
Merge branch 'develop' of github.com:Medicare-Call/Medicare-Call-Android into feat/fcm-alarm-#110

# Conflicts:
#	app/src/main/java/com/konkuk/medicarecall/MainActivity.kt

alswlekk committed:
Feat: SharedPreferences에 저장하는 형식에서 DataStore(직렬화)로 통일

alswlekk committed:
Merge remote-tracking branch 'origin/feat/fcm-alarm-#110' into feat/fcm-alarm-#110

# Conflicts:
#	app/src/main/java/com/konkuk/medicarecall/data/api/AppPreferences.kt
#	app/src/main/java/com/konkuk/medicarecall/data/api/fcm/FcmService.kt

alswlekk committed:
Merge remote-tracking branch 'origin/feat/fcm-alarm-#110' into feat/fcm-alarm-#110

# Conflicts:
#	app/src/main/java/com/konkuk/medicarecall/data/api/AppPreferences.kt
#	app/src/main/java/com/konkuk/medicarecall/data/api/fcm/FcmService.kt

alswlekk committed:
Merge remote-tracking branch 'origin/feat/fcm-alarm-#110' into feat/fcm-alarm-#110

alswlekk committed:
Refactor: dataStore 이용해서 저장하는 형식으로 수정

alswlekk committed:
Refactor: dataStore 이용해서 저장하는 형식으로 수정

alswlekk committed:
Merge remote-tracking branch 'origin/feat/fcm-alarm-#110' into feat/fcm-alarm-#110

alswlekk committed:
Merge branch 'develop' into feat/fcm-alarm-#110

alswlekk committed:
Refactor: MainActivity 파일 수정(중복 실행 방지)

alswlekk committed:
Merge remote-tracking branch 'origin/feat/fcm-alarm-#110' into feat/fcm-alarm-#110

alswlekk committed:
Merge remote-tracking branch 'origin/feat/fcm-alarm-#110' into feat/fcm-alarm-#110

alswlekk committed:
Merge remote-tracking branch 'origin/feat/fcm-alarm-#110' into feat/fcm-alarm-#110

alswlekk committed:
Refactor: CI 관련 오류 코드 수정

alswlekk committed:
Refactor: CI 관련 오류 코드 수정

alswlekk committed:
Refactor: CI 관련 오류 코드 수정

alswlekk committed:
Merge remote-tracking branch 'origin/feat/fcm-alarm-#110' into feat/fcm-alarm-#110

alswlekk committed:
Refactor: CI 관련 오류 코드 수정

alswlekk committed:
Refactor: CI 관련 오류 코드 수정

alswlekk committed:
Merge remote-tracking branch 'origin/feat/fcm-alarm-#110' into feat/fcm-alarm-#110

alswlekk committed:
Refactor: CI 관련 오류 코드 수정

alswlekk committed:
Refactor: CI 관련 오류 코드 수정

alswlekk committed:
Merge remote-tracking branch 'origin/feat/fcm-alarm-#110' into feat/fcm-alarm-#110

alswlekk committed:
Merge remote-tracking branch 'origin/feat/fcm-alarm-#110' into feat/fcm-alarm-#110

alswlekk committed:
Merge remote-tracking branch 'origin/feat/fcm-alarm-#110' into feat/fcm-alarm-#110

alswlekk committed:
Refactor: Fcm 관련 코드 기존의 토큰 관련 코드랑 독립적으로 분리

alswlekk committed:
Refactor: Fcm 관련 코드 기존의 토큰 관련 코드랑 독립적으로 분리

alswlekk committed:
Merge remote-tracking branch 'origin/feat/fcm-alarm-#110' into feat/fcm-alarm-#110

alswlekk committed:
Merge branch 'develop' of github.com:Medicare-Call/Medicare-Call-Android into feat/fcm-alarm-#110

Copilot committed:
Initial plan

alswlekk committed:
Refactor: CI 관련 수정사항 반영

alswlekk committed:
Refactor: CI 관련 수정사항 반영

Copilot committed:
Initial setup: Copy files from FCM branch for fixes

Co-authored-by: alswlekk <162430621+alswlekk@users.noreply.github.com>

Copilot committed:
fix(detekt): remove extra blank line, fix FcmService indentation, remove unused import

Co-authored-by: alswlekk <162430621+alswlekk@users.noreply.github.com>

alswlekk committed:
Merge branch 'feat/fcm-alarm-#110' into copilot/fix-detekt-style-issues

alswlekk committed:
Merge pull request #170 from Medicare-Call/copilot/fix-detekt-style-issues

Fix Detekt style violations: indentation, blank lines, unused import

alswlekk committed:
Merge remote-tracking branch 'origin/feat/fcm-alarm-#110' into feat/fcm-alarm-#110

alswlekk committed:
Refactor: CI 관련 수정사항 반영

ProtossManse committed:
Fix: #110 잘못된 baseurl 키 변경

ProtossManse commented:
LGTM

alswlekk committed:
Merge branch 'develop' of github.com:Medicare-Call/Medicare-Call-Android into feat/fcm-alarm-#110

alswlekk committed:
Fix: 서버와 FCM 관련해 로그인, 회원가입 API 연동

alswlekk committed:
Fix: ci 관련 코드 수정
